torch/csrc/jit/serialization/unpickler.cpp

```diff
@@ -647,6 +651,35 @@ void Unpickler::readGlobal(
   stack_.emplace_back(int64_t(globals_.size() - 1));
 }
 
+void Unpickler::rebuildSparseCooTensor() {
+  globals_.emplace_back([this] {
+    auto tup = pop(stack_).toTuple();
+    const auto& elements = tup->elements();
+    size_t idx = 0;
+    std::vector<int64_t> size = tupleToIntList(elements.at(idx++));
+    bool requires_grad = elements.at(idx++).toBool();
+    bool pinned_memory = elements.at(idx++).toBool();
+    c10::ScalarType data_type =
+        static_cast<c10::ScalarType>(elements.at(idx++).toInt());
+    c10::Device device(elements.at(idx++).toStringRef());
+    auto options = c10::TensorOptions()
+                       .dtype(data_type)
+                       .layout(c10::Layout::Sparse)
+                       .device(device)
+                       .pinned_memory(pinned_memory)
```

Comment:
When rebuilding the dense tensor it uses the TensorOptions from the storage `      result = at::empty({0}, storage_tensor.options());
`.

Since sparse tensors do not have storage, I pushed the values required to build TensorOptions. TensorOptions is used when building the result sparse tensor.

Reply:
> Is this same as sparse_tensor.values().options().layout(c10::Layout::Sparse)?
What do you mean?

This is for setting the options of the sparse tensor, not the values or indices.

Reply:
> What do you mean?

Yep, I am curious whether these two will return the same options value?

1. in comment
```c++
auto options = values_tensor.options().layout(c10::Layout::Sparse);
```

2.  in PR
```c++
bool requires_grad = elements.at(idx++).toBool();
bool pinned_memory = elements.at(idx++).toBool();
c10::ScalarType data_type =
        static_cast<c10::ScalarType>(elements.at(idx++).toInt());
auto options = c10::TensorOptions()
                       .dtype(data_type)
                       .layout(c10::Layout::Sparse)
                       .pinned_memory(pinned_memory)
                       .requires_grad(requires_grad);
```